### ISSUE TITLE
inaccurate marker positioning in `plot()`

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#7233]
mpl 1.5.3/2.0b4/master, no matplotlibrc
```
_xs = np.linspace(0, 30, 201); plt.plot(_xs, scipy.stats.lognorm.pdf(_xs / 14, .472) / 14, ".")
```
![plot](https://cloud.githubusercontent.com/assets/1322974/19205573/2ae8b19c-8c98-11e6-84f8-89ecbdca82c1.png)

Note the slight inaccuracy in positioning of the markers at the positioning of the markers at the positions marked with the red circles, resulting in a "step-like" behavior.

The data itself does not present this behavior, as can be checked just by zooming in these regions.

**Comment by efiring** [comment]
I wonder whether it is just a limitation on the ability of antialiasing to deliver a visual impression of sub-pixel accuracy.  The steppiness is at the one-pixel level, but it doesn't look like simple pixel-snapping.

**Comment by tacaswell** [comment]
Turning snapping off does not help that much.  I wonder if this is related to the way we 'stamp' the markers?

attn @mdboom

**Comment by tacaswell** [comment]
```python
import numpy
import scipy.stats
import matplotlib.pyplot as plt
plt.ion()
fig, ax = plt.subplots()
_xs = np.linspace(0, 30, 201); 
ln, = ax.plot(_xs, scipy.stats.lognorm.pdf(_xs / 14, .472) / 14, ".")
```

**Comment by mdboom** [comment]
Yes, it's definitely related to how we stamp the markers.  We don't do individual subpixel positioning for the markers.  We could do that but it would be significantly slower (so should be a user-configurable option).

**Comment by gson1703** [comment]
I'm also suffering from this issue.  See the attached script and image for an example of what happens to a grid of evenly spaced markers.  This level of quality is simply not acceptable for professional use - it looks like something out of the 1980s, but this 2017, and subpixel positioning is no longer optional.

[grid.py.txt](https://github.com/matplotlib/matplotlib/files/1197011/grid.py.txt)
![detail](https://user-images.githubusercontent.com/1098450/28920091-028678b2-7859-11e7-88a2-7af64754a672.png)

**Comment by WeatherGod** [comment]
Could you repost that example. The link appears to be broken.

**Comment by WeatherGod** [comment]
I would be curious if this is something that could be resolved with the Cairo rework?

**Comment by gson1703** [comment]
```python
import math
import numpy
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
plt.style.use('dark_background')
fig = plt.figure()
ax = fig.add_subplot(111)
xv = []
yv = []
for y in range(10):
    for x in range(10):
        xv.append(x * 0.147)
        yv.append(y * 0.147)
ax.set_xlim(0, 10)
ax.set_ylim(0, 10)
plt.scatter(xv, yv)
fig.savefig('grid.png')
```

**Comment by anntzer** [comment]
@WeatherGod It is indeed solved by mpl_cairo (in fact this was one of the original motivations for writing it).  See second example at https://gitter.im/matplotlib/matplotlib?at=5963202789aea4761d81a829.  (You even commenter right under it :-))
As a side note mpl_cairo also stamps the markers much faster (there's a timing script in the examples).

**Comment by WeatherGod** [comment]
I hadn't noticed the marker positioning examples. I was more interested in
the text kerning at the time. That's great to know that the Cairo-based
backends can be a suitable work-around when they are available (I am open
to making them the default over AGG, but that might have to wait until
v3.0, and wider adoption of the newer C++ standards across compilers).

**Comment by anntzer** [comment]
I'm sure you can technically make it work in Agg... if someone actually understands the Agg rendering code paths (I gave up due to complete lack of documentation on Agg's side).  The relevant part, which is at https://github.com/anntzer/mpl_cairo/blob/master/src/_mpl_cairo.cpp#L540 (cache preparation) and https://github.com/anntzer/mpl_cairo/blob/master/src/_mpl_cairo.cpp#L588 (marker stamping), should be fairly self explanatory if you know C++ -- I make a cache of renderered markers with various subpixel offsets (decided according to the path.simplify_theshold, because why not) and then the stamping loop just needs to pull out the right one each time.

There's some discussion about using a recent g++ on a manylinux1 docker, which also involves statically linking a recent libstdc++ into a manylinux1 wheel, so that's apparently doable too... but tbh it's not very high on my priority list now.

**Comment by ImportanceOfBeingErnest** [comment]
There is a nice visualization of this issue brought up in [this Stackoverflow question](https://stackoverflow.com/questions/48979691/sub-pixel-accuracy-scatter-plots-with-matplotlib). 

Running this with TkCairo shows the same issue as with TkAgg in the saved animation.  
Interestingly the scatter animates smoothly on screen with Cairo backend, while there is no difference between the animation on screen and the saved gif for Agg backend. (@anntzer Is this difference to be expected?)


![skewedgridtkcairo](https://user-images.githubusercontent.com/23121882/37561868-0ace5e64-2a5a-11e8-882d-8a13f2df1e62.gif)

Code for reproduction:

```
import numpy as np
import matplotlib
matplotlib.use("TkCairo")
import matplotlib.pyplot as plt
from matplotlib.collections import PatchCollection
from matplotlib.animation import FuncAnimation

x, y = np.meshgrid(np.linspace(-1,1,31),np.linspace(-1,1,31))

fig, (ax1,ax2) = plt.subplots(ncols=2, figsize=(5,2.8))
fig.subplots_adjust(.02,.02,.98,.9,.03,.03)
fig.suptitle(matplotlib.get_backend())
ax1.set_title("scatter")
ax2.set_title("PatchCollection")
### Scatter
sc = ax1.scatter(x,y,s=2)
### PatchCollection
patches=[]
for i in range(x.shape[0]):
    for j in range(x.shape[1]):
        patches.append(plt.Circle((x[i,j],y[i,j]),0.015))
        
class UpdatablePatchCollection(PatchCollection):
    def __init__(self, patches, *args, **kwargs):
        self.patches = patches
        PatchCollection.__init__(self, patches, *args, **kwargs)
    def get_paths(self):
        self.set_paths(self.patches)
        return self._paths
    
pc = UpdatablePatchCollection(patches)
ax2.add_collection(pc)


for ax in (ax1,ax2):
    ax.axis("off")
    ax.set_xlim(-1.2,1.2)
    ax.set_ylim(-1.2,1.2)
    ax.set_aspect('equal')

def generator():
    X=np.copy(x); Y=np.copy(y)
    for i in range(100):
        X += 0.0015*np.cos(x+3*y)
        Y += 0.0015*np.sin(6*x-4*y)
        yield (X,Y)

def animate(i): 
    x,y = i
    # update scatter
    sc.set_offsets(np.c_[x.flatten(),y.flatten()])
    # update PatchCollection
    for circle, xi,yi in zip(patches, x.flatten(),y.flatten()):
        circle.center = (xi,yi)
    pc.pchanged()

ani = FuncAnimation(fig, animate, frames=generator, interval=50, repeat=False)
ani.save("skewedgrid{}.gif".format(matplotlib.get_backend()), writer="imagemagick")
plt.show()
```

**Comment by anntzer** [comment]
The difference is due to the fact that when saving an animation, we internally fallback on agg, because backend_cairo did not implement a way to dump the raw buffer.  https://github.com/matplotlib/matplotlib/pull/10828 fixes that issue.

**Comment by gokceneraslan** [comment]
Just to make sure, the reason for these weird horizontal and vertical patterns below is due to this issue basically, right?

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.tile(np.arange(100), 100)
y = np.repeat(np.arange(100), 100)

f, ax = plt.subplots(figsize=(7, 7))
ax.scatter(x, y, s=1, c='k')
```

![image](https://user-images.githubusercontent.com/1140359/50424132-62769580-082d-11e9-99eb-872935df054e.png)

**Comment by anntzer** [comment]
Yes.

**Comment by jklymak** [comment]
Is there really an action item for this?  If not, I'll leave open but un-milestone.

**Comment by tacaswell** [comment]
Yes, instead of generating 1 stamp and snapping to the nearest pixel, generate N and snap to nearest fraction of a pixel.

**Comment by Jona-L** [comment]
not sure if related or said before, but the pixel marker position is too high when other markers are fine
`   plt.xlim(0, 3)
    plt.ylim(0, 3)
    plt.plot(.05, 1, ',b')
    plt.plot(.1, 1, '.b')
    plt.plot(.2, 1, 'ob', ms=10, mew=1)
    plt.plot(.3, 1, ',b')
    plt.plot(.3, 1, 'or', fillstyle='none', ms=10, mew=1)
    plt.plot(.4, 1, '.b', ms=9)
    plt.plot(.4, 1, 'or', fillstyle='none', ms=10, mew=1)    
    plt.savefig('tst.png', dpi=100) `
![image](https://user-images.githubusercontent.com/14801497/79998119-92ce5280-84ba-11ea-8103-e7ac46e24326.png)
